File: python/string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        s = s.strip()
        if len(s) == 0: return 0
        if s[0] == "-":
            sign = -1
            s = s[1:]
        elif s[0] == "+":
            sign = 1
            s = s[1:]
        else:
            sign = 1
        
        res = 0
        i = 0
        while i < len(s) and s[i].isdigit():
            res *= 10
            res += int(s[i])
            i += 1
        res = res * sign
        if res < -2**31:
            return -2**31
        elif res > 2**31-1:
            return 2**31-1
        else:
            return res

File: go/string_to_integer_atoi.go
package leetcode

import (
	"math"
	"strings"
)

func myAtoi(s string) int {
	s = strings.Trim(s, " ")
	sRunes := []rune(s)

	if len(sRunes) == 0 {
		return 0
	}
	var sign int
	if sRunes[0] == '-' {
		sign = -1
		sRunes = sRunes[1:]
	} else if sRunes[0] == '+' {
		sign = 1
		sRunes = sRunes[1:]
	} else {
		sign = 1
	}

	res := 0
	Upper, Lower := int(math.Pow(2, 31))-1, -int(math.Pow(2, 31))
	for i := 0; i < len(sRunes) && sRunes[i] >= 48 && sRunes[i] <= 57; i++ {
		res *= 10
		res += int(sRunes[i]) - 48
		if res*sign < Lower {
			return Lower
		} else if res*sign > Upper {
			return Upper
		}
	}

	res *= sign
	if res < int(-math.Pow(2, 31)) {
		return -int(math.Pow(2, 31))
	} else if res > int(math.Pow(2, 31))-1 {
		return int(math.Pow(2, 31)) - 1
	} else {
		return res
	}
}
